applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/eigenfrequency_response_function_utility.h

```diff
@@ -314,33 +315,44 @@ class EigenfrequencyResponseFunctionUtility
                         gradient_contribution[coord_dir_i] += gradient_prefactors[i] * inner_prod(eigenvectors_of_element[i] , aux_vector) * mWeightingFactors[i];
                     }
 
-                   node_i.GetInitialPosition()[coord_dir_i] -= mDelta;
+                    node_i.GetInitialPosition()[coord_dir_i] -= mDelta;
+                    node_i.Coordinates()[coord_dir_i] -= mDelta;
                 }
                 noalias(node_i.FastGetSolutionStepValue(SHAPE_SENSITIVITY)) += gradient_contribution;
             }
         }
     }
 
     // --------------------------------------------------------------------------
-    double GetEigenvalue(int eigenfrequency_id)
+    double GetEigenvalue(const int& eigenfrequency_id)
     {
         return (mrModelPart.GetProcessInfo()[EIGENVALUE_VECTOR])[eigenfrequency_id-1];
     }
 
     // --------------------------------------------------------------------------
-    void DetermineEigenvectorOfElement(ModelPart::ElementType& traced_element, int eigenfrequency_id, int size_of_eigenvector, Vector& rEigenvectorOfElement)
+    void DetermineEigenvectorOfElement(ModelPart::ElementType& rElement, const int& eigenfrequency_id, Vector& rEigenvectorOfElement, ProcessInfo& CurrentProcessInfo)
     {
-        rEigenvectorOfElement.resize(size_of_eigenvector,false);
+        const std::size_t num_nodes = rElement.GetGeometry().size();
 
-        const std::size_t num_nodes = traced_element.GetGeometry().size();
-        const std::size_t num_node_dofs = size_of_eigenvector/num_nodes;
+        std::vector<std::size_t> eq_ids;
+        rElement.EquationIdVector(eq_ids, CurrentProcessInfo);
 
-        for (std::size_t node_index=0; node_index<num_nodes; node_index++)
+        if (rEigenvectorOfElement.size() != eq_ids.size())
         {
-            Matrix& rNodeEigenvectors = traced_element.GetGeometry()[node_index].GetValue(EIGENVECTOR_MATRIX);
+            rEigenvectorOfElement.resize(eq_ids.size(), false);
+        }
+
+        for (auto& r_node_i : rElement.GetGeometry())
+        {
+            const ModelPart::NodeType::DofsContainerType& r_node_dofs = r_node_i.GetDofs();
 
-            for (std::size_t dof_index = 0; dof_index < num_node_dofs; dof_index++)
-                rEigenvectorOfElement(dof_index+num_node_dofs*node_index) = rNodeEigenvectors((eigenfrequency_id-1),dof_index);
+            const Matrix& rNodeEigenvectors = r_node_i.GetValue(EIGENVECTOR_MATRIX);
+            for (std::size_t dof_index = 0; dof_index < r_node_dofs.size(); dof_index++)
+            {
+                auto it_dof = std::begin(r_node_dofs) + dof_index;
+                const std::size_t elem_dof_index = std::distance(eq_ids.begin(), std::find(eq_ids.begin(), eq_ids.end(), it_dof->EquationId()));
```

Comment:
I don't really like this line, but the only alternative i could come up with is to create an intermediate map to do the search, and that seems to be slower.

Reply:
I don't fully understand what you are doing here, can you explain a bit pls?

Reply:
The eigenvectors are stored at the nodes in the `EIGENVECTOR_MATRIX`.
I need the part of the full eigenvector, that belongs to a specific element, in the order of the element dofs (because i multiply with the elemental stiffness matrix...)

The components of the `EIGENVECTOR_MATRIX` are ordered according to the Dof list of the node. The ordering of the dofs at the element and may not be the same. This is why i need to compare the equation ids of the dof lists coming from the node and the element.

e.g. for shells:

* dofs at node: [Rotation, Displacement]
* dofs at element (for each node): [Displacement, Rotation]

Reply:
hm ok, this i a bit over my head, maybe @msandre knows a better solution...?